### ISSUE TITLE
npm run build.prod.exp fails

### ISSUE THREAD

**Issue by lukefarran** [NathanWalker/angular-seed-advanced#224]
**IMPORTANT**: This repository's issues are reserved for feature requests and bug reports. Do not submit support requests here.

**Steps to reproduce and a minimal demo of the problem**
run 'npm run build.prod.exp' with commit 7042bea3d54c673fa9c21df6bf90b1e64593032a

Generates the following exception at the gulp 'compile.ahead.prod' task:
Error: Error encountered resolving symbol values statically. Function calls are not supported. Consider replacing the function or lambda with a reference to an exported function (position 28:19 in the original .ts file), resolving symbol MultilingualModule in /Users/lukefarran/ProjectsGithub/angular2-seed-advanced/dist/tmp/app/frameworks/i18n/multilingual.module.ts
    at simplifyInContext (/Users/lukefarran/ProjectsGithub/angular2-seed-advanced/node_modules/@angular/compiler-cli/src/static_reflector.js:473:23)
    at StaticReflector.simplify (/Users/lukefarran/ProjectsGithub/angular2-seed-advanced/node_modules/@angular/compiler-cli/src/static_reflector.js:476:22)
    at StaticReflector.annotations (/Users/lukefarran/ProjectsGithub/angular2-seed-advanced/node_modules/@angular/compiler-cli/src/static_reflector.js:61:36)
    at _loop_1 (/Users/lukefarran/ProjectsGithub/angular2-seed-advanced/node_modules/@angular/compiler-cli/src/codegen.js:53:54)
    at CodeGenerator.readFileMetadata (/Users/lukefarran/ProjectsGithub/angular2-seed-advanced/node_modules/@angular/compiler-cli/src/codegen.js:66:13)
    at /Users/lukefarran/ProjectsGithub/angular2-seed-advanced/node_modules/@angular/compiler-cli/src/codegen.js:100:74
    at Array.map (native)
    at CodeGenerator.codegen (/Users/lukefarran/ProjectsGithub/angular2-seed-advanced/node_modules/@angular/compiler-cli/src/codegen.js:100:35)
    at codegen (/Users/lukefarran/ProjectsGithub/angular2-seed-advanced/tools/tasks/seed/compile.ahead.prod.js:10:86)
    at Object.main (/Users/lukefarran/ProjectsGithub/angular2-seed-advanced/node_modules/@angular/tsc-wrapped/src/main.js:30:16)
Compilation failed

_What steps should we try in your demo to see the problem?_

**Current behavior**
Fails to complete 'npm run build.prod.exp'

**Expected/desired behavior**
Successfully complete.  This works in the parent angular2-seed

**Other information**

**Comment by NathanWalker** [comment]
`exp` == `experimental` ;)

If anyone can help look, would be appreciated. Will be 2 weeks before I get a chance.

**Comment by lfarran** [comment]
@NathanWalker I'll try to help out as I've been using AoT.  It's picky and tricky, hopefully it'll be improved in the next RC or official release.

**Comment by kbrandwijk** [comment]
There is a pull request on ng2-translate to fix this: [https://github.com/ocombe/ng2-translate/pull/229](https://github.com/ocombe/ng2-translate/pull/229). Best to wait for it.

**Comment by kbrandwijk** [comment]
And the pull request was done. Now trying with ng2-translate 2.4.4. Seems to have fixed the problem. However, it's just the first one of a longer list of errors that will appear once this one is fixed. I'll try to prepare a PR.

**Comment by NathanWalker** [comment]
Thanks @BrainsoftLtd ! Lemme know. 👍

**Comment by fbascheper** [comment]
I still see errors with build.prod.exp.

```
Error: Error encountered resolving symbol values statically. Function calls are not supported. Consider replacing the function or lambda with a reference to an exported function (position 28:19 in the original .ts file), resolving symbol MultilingualModule in /klant/connectender/website/ng2-connectender/dist/tmp/app/frameworks/i18n/multilingual.module.ts
    at simplifyInContext (/Users/iminar/Dev/angular/modules/@angular/compiler-cli/src/static_reflector.ts:528:15)
    at StaticReflector.simplify (/Users/iminar/Dev/angular/modules/@angular/compiler-cli/src/static_reflector.ts:532:18)
    at StaticReflector.annotations (/Users/iminar/Dev/angular/modules/@angular/compiler-cli/src/static_reflector.ts:92:28)
    at _loop_1 (/Users/iminar/Dev/angular/modules/@angular/compiler-cli/src/codegen.ts:63:48)
    at CodeGenerator.readFileMetadata (/Users/iminar/Dev/angular/modules/@angular/compiler-cli/src/codegen.ts:57:22)
    at /Users/iminar/Dev/angular/modules/@angular/compiler-cli/src/codegen.ts:105:56
    at Array.map (native)
    at CodeGenerator.codegen (/Users/iminar/Dev/angular/modules/@angular/compiler-cli/src/codegen.ts:105:33)
    at codegen (/klant/connectender/website/ng2-connectender/tools/tasks/seed/compile.ahead.prod.ts:14:69)
    at Object.main (/klant/connectender/website/tools/@angular/tsc-wrapped/src/main.ts:44:12)
Compilation failed
```

This is fixed by exporting the function like so

```
export function newTranslateLoader(http: Http) {
  return new TranslateStaticLoader(http, 'assets/i18n', '.json');
}

@NgModule({
  imports: [
    CommonModule,
    RouterModule,
    FormsModule,
    HttpModule,
    TranslateModule.forRoot({
      provide: TranslateLoader,
      deps: [Http],
      useFactory: newTranslateLoader
    })
  ],
  declarations: [
```

However, this raises another issue, namely [Angular #11307 ](https://github.com/angular/angular/issues/11307)

**Comment by shripalsoni04** [comment]
I removed all the components/frameworks code and created just one component to see if AOT works or not. 

But, after doing that `'compile.ahead.prod'` task is throwing below error:
`Error: Unexpected value 'AppComponent' declared by the module 'WebModule'`

I tried to found the root cause of the issue and found this comment https://github.com/angular/angular/issues/7968#issuecomment-223779992. So I think our custom decorator `BaseComponent` will not work with AOT as per that comment.

When I replaced `BaseComponent` to `Component` as shown below, I was able to build the project successfully.

```
@Component({
  moduleId: module.id,
  selector: 'sd-app',
  templateUrl: 'app.component.html',
  changeDetection: ChangeDetectionStrategy.Default // Everything else uses OnPush
})
export class AppComponent {
...
}
```

But this will remove the purpose of using same component across platforms. I think to make it work with AOT, we need to have different components for platforms which requires different metadata (i.e. templateUrl) and share code using services. Let me know if anyone has any other simple solution for this.

Thanks.
